core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java

```diff
@@ -318,4 +423,46 @@ public synchronized T get()
             }
         }
     }
+
+    private static Set<Long> createSetForLongJavaType(Set<Long> values, Type type, MethodHandle hashCodeHandle, MethodHandle equalsHandle)
+    {
+        if (values.isEmpty() || !isDirectLongComparisonValidType(type)) {
+            return new LongOpenCustomHashSet(values, 0.25f, new LongStrategy(hashCodeHandle, equalsHandle, type));
+        }
+
+        long min = Long.MAX_VALUE;
+        long max = Long.MIN_VALUE;
+        for (Long value : values) {
+            min = min(min, value);
+            max = max(max, value);
+        }
+
+        // A Set based on a fastutil open hashset uses (next power of two of size / load-factor)
+        // slots in a hash table which is an array of longs.
+        long setEntries = (int) min(1 << 30, nextPowerOfTwo((long) ceil(values.size() / 0.25f)));
+        BigInteger range = BigInteger.valueOf(max)
+                .subtract(BigInteger.valueOf(min))
+                .add(BigInteger.valueOf(1));
+        // A Set based on a bitmap uses (max - min) / 64 longs
+        // Create a bitset only if it uses fewer entries than the equivalent hash set
+        if (range.compareTo(BigInteger.valueOf(Integer.MAX_VALUE)) > 0
+                || (range.intValueExact() / 64) + 1 > setEntries) {
+            return new LongOpenHashSet(values, 0.25f);
+        }
+        return new LongBitSetFilter(values, min, max);
+    }
+
+    private static boolean isDirectLongComparisonValidType(Type type)
```

Comment:
We have this also in `SimplifyContinuousInValues`. I would move it to `io.trino.type.TypeUtils`

Reply:
That method is slightly different, over there we're additionally looking for types where it's possible to get the next consecutive value by just incrementing the underlying long by 1.
I've added a code comment and renamed it to be clearer.